pkg/collection/ducks.go

```diff
@@ -0,0 +1,200 @@
+}
+
+type DuckFilters struct {
+	// DuckLabel is expected to be in the form:
+	//	 `<group>/<names.singular>`
+	DuckLabel string
+	// DuckVersionFormat is expected to be in the form:
+	//   `<names.plural>.<group>`
+	// and will be used to assemble `<names.plural>.<group>/<duckVersion>`.
+	// This will group duck type versions based on the given annotation base.
+	DuckVersionPrefix string
+}
+
+// NewDuckHunter
+// versions are used to default all the DuckType versions that apply to unfiltered CRDs.
+func NewDuckHunter(versions []v1alpha1.DuckVersion, filters *DuckFilters) DuckHunter {
+	dh := &duckHunter{
+		filters:  filters,
+		versions: make([]string, 0),
+		ducks:    make(map[string][]v1alpha1.ResourceMeta, len(versions)),
+	}
+
+	for _, v := range versions {
+		if _, found := dh.ducks[v.Name]; !found {
+			dh.versions = append(dh.versions, v.Name)
+			dh.ducks[v.Name] = make([]v1alpha1.ResourceMeta, 0)
+		}
+	}
+
+	return dh
+}
+
+type duckHunter struct {
+	filters *DuckFilters
+	// versions are the versions to use for unfiltered CRDs being added to the ducks map.
+	versions []string
+	ducks    map[string][]v1alpha1.ResourceMeta
+}
+
+func (dh *duckHunter) AddCRDs(crds []*apiextensionsv1.CustomResourceDefinition) {
+	for _, crd := range crds {
+		dh.AddCRD(crd)
+	}
+}
+
+func (dh *duckHunter) AddCRD(crd *apiextensionsv1.CustomResourceDefinition) {
+	if crd == nil {
+		return
+	}
+	if metas := crdToResourceMeta(crd); len(metas) > 0 {
+		dh.collectVersionsByFilter(crd)
+		for _, meta := range metas {
+			if !dh.handledWithFilters(crd, meta) {
+				// TODO: here is where the label filters would be tested for duck version annotation matching.
+				for _, v := range dh.versions {
+					dh.ducks[v] = append(dh.ducks[v], meta)
+				}
+			}
+		}
+	}
+}
+
+func (dh *duckHunter) handledWithFilters(crd *apiextensionsv1.CustomResourceDefinition, meta v1alpha1.ResourceMeta) bool {
+	if dh.filters == nil {
+		return false
+	}
+	for k, v := range crd.Labels {
+		if dh.filters.DuckLabel != "" {
+			if k == dh.filters.DuckLabel {
+				if v == "true" {
+					return dh.insertHandledDuckByVersionFilter(crd, meta)
+				}
+				return true
+			}
+		} else {
+			return dh.insertHandledDuckByVersionFilter(crd, meta)
+		}
+	}
+	return false
+}
+
+// this makes sure that there is a slice for each found version of the duck.
+func (dh *duckHunter) collectVersionsByFilter(crd *apiextensionsv1.CustomResourceDefinition) {
+	if dh.filters == nil || dh.filters.DuckVersionPrefix == "" {
+		return
+	}
+	for k := range crd.Annotations {
+		if strings.HasPrefix(k, dh.filters.DuckVersionPrefix+"/") {
+			version := strings.TrimPrefix(k, dh.filters.DuckVersionPrefix+"/")
+			if _, found := dh.ducks[version]; !found {
+				dh.ducks[version] = make([]v1alpha1.ResourceMeta, 0)
+			}
+		}
+	}
+}
+
+func (dh *duckHunter) insertHandledDuckByVersionFilter(crd *apiextensionsv1.CustomResourceDefinition, meta v1alpha1.ResourceMeta) (handled bool) {
```

Comment:
nit: do you think is worth having a named return here? 
I think we haven't been using this throughout the code, and I'd rather be more consistent. In general, I don't like them much, but if you want to keep it, I'm also fine...

Reply:
in this case the named return does make sense only because to create a tracking variable and return it is the same difference, if you think it is too unreadable, I can change it. For an internal method that is tightly scoped in task it felt ok to me

Reply:
fine to keep it I guess..